[50954] Record progress in multiple units

cbliard opened the pull request:
https://community.openproject.org/wp/50954

aaron-contreras committed:
Add `gitlab_chronic_duration`

See: https://gitlab.com/gitlab-org/ruby/gems/gitlab-chronic-duration

Really nice gem to handle parsing and outputting durations in multiple
units.

aaron-contreras committed:
Parse and format modal progress fields

aaron-contreras committed:
Extract parsing and formatting logic for duration into service

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Display chronic durations on all work package contexts

aaron-contreras committed:
Format journal entries chronically for work/remaining work

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras commented:
eslint is complaining about the `chronic_duration.js` file. This is taken from Gitlab's implementation and is meant to be kept inline with their `chronic_duration` gem. There's no npm package for `chronic_duration.js` as it's directly part of the gitlab source code. Would vendoring the file prevent eslint from trying to check it or is there a config to exclude this file from eslint. Probably has been done before in the core repo but unsure of how to handle it @oliverguenther

aaron-contreras committed:
Adjust specs to new duration formats

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Fix `bin/setup_dev` script

Missing dependency in `package-lock.json`

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Ignore chronic_duration.js from eslint

This file is sourced from Gitlab's Rails application and provided as
`.js`. We therefore don't want to enforce eslint rules on it to keep
it prisitine with its upstream source.

aaron-contreras committed:
Fix eslint complaint about 'eqeqeq' rule

aaron-contreras committed:
Rename WorkingDays constant and route to WorkingDaysAndHours

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras commented:
@cbliard 'eqeqeq' eslint issue addressed more globally in https://github.com/opf/openproject/pull/15560

aaron-contreras committed:
Make hours/day, hours/week, days/month configurable

aaron-contreras committed:
Provide duration configuration to frontend

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Sync days_per_week and days_per_month on client side

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Ensure backend validation for days_per_week/days_per_month consistency

aaron-contreras committed:
Adapt specs to new format

aaron-contreras committed:
Lint ABC branch condition size

cbliard reviewed (commented):
I played a little with it. It responds really well. Good job!
3 things I wonder about, and probably @psatyal can help us on it:

- what do we do with text longer than the allowed size? Currently it's clipping. Do we extend the default width?
- what about the precision? When entering 1m, it can't represent it accurately with only 2 decimals. It results to "1m 12s". Not a big deal, but it's kinda surprising to enter "2h 12m" and end up with "2h 12m 0.0000000001s", or enter "2h 13m" and end up with "2h 13m 12.00000000001s". Should we switch from floating points holding nb of hours to integer holding nb of seconds?
- when omitting the "m", or before entering it, the value is interpreted as hours. It would be better if interpreted as minutes. This is especially visible when entering a value for the first time: with w -> rw, and typing each character one by one, it gives this:
    - work -> remaining work (when 100%)
    - 2 -> 2h
    - 2h -> 2h
    - 2h1 -> 3h (would have been expecting 2h 1m)
    - 2h15 -> 17h (would have been expecting 2h 15m)
    - 2h15m -> 2h 15m

cbliard commented on app/controllers/admin/settings/working_days_and_hours_settings_controller.rb:
```diff
@@ -64,6 +65,11 @@ def non_working_days_params
+      params.require(:settings)
+            .permit(:hours_per_day, :days_per_week, :days_per_month)
+    end
+
```
The fact that `#non_working_days_params` and `#duration_configuration_params` are both doing `params.require` with different arguments leads to a "Unpermitted params" everytime which feels like an error on the console:
```
D, [2024-05-24T11:21:49.662215 #70386] DEBUG -- : Unpermitted parameters: :hours_per_day, :days_per_week, :days_per_month, :working_days. Context: { controller: Admin::Settings::WorkingDaysAndHoursSettingsController, action: update, request: #<ActionDispatch::Request:0x000000033fcac820>, params: {"utf8"=>"✓", "_method"=>"patch", "authenticity_token"=>"[FILTERED]", "settings"=>{"hours_per_day"=>"abcd", "days_per_week"=>"50", "days_per_month"=>"200.5", "working_days"=>["", "1", "2", "3", "4", "5"]}, "Non-working day"=>"", "controller"=>"admin/settings/working_days_and_hours_settings", "action"=>"update"} }
D, [2024-05-24T11:21:49.662389 #70386] DEBUG -- : Unpermitted parameter: :working_days. Context: { controller: Admin::Settings::WorkingDaysAndHoursSettingsController, action: update, request: #<ActionDispatch::Request:0x000000033fcac820>, params: {"utf8"=>"✓", "_method"=>"patch", "authenticity_token"=>"[FILTERED]", "settings"=>{"hours_per_day"=>"abcd", "days_per_week"=>"50", "days_per_month"=>"200.5", "working_days"=>["", "1", "2", "3", "4", "5"]}, "Non-working day"=>"", "controller"=>"admin/settings/working_days_and_hours_settings", "action"=>"update"} }
```

Actually, the permitted params are given by `settings_params` from the super class which is doing `permitted_params.settings.to_h` which uses settings definitions to get all allowed keys. The keys `:hours_per_day, :days_per_week, :days_per_month` are already part of the `settings` being returned in the parent method. So it should not be needed to call `duration_configuration_params` at all.

Yep, I just tried and it works.

There is still one logged error line, but this one was already there before.

cbliard commented on app/contracts/settings/working_days_and_hours_params_contract.rb:
```diff
@@ -0,0 +1,96 @@
+    validate :hours_per_day_are_present
+    validate :days_per_week_are_present
+    validate :days_per_month_are_present
+    validate :days_per_week_and_days_per_month_are_consistent
```
It does not validate enough IMHO:
- It should not be possible to set more than 24h per day, 7 days per week, or 31 days per month. It should raise an error.
- It should not be possible to enter negative values
- It should not be possible to enter zero values or text. Using text is "blocked" client side, but not checked server side. When doing so with a curl request, the value is interpreted as 0 due to some `to_i` or `to_f` here and there.
  - When hours per day value is 0, it leads to division by zero errors when converting time

cbliard commented on app/views/admin/settings/working_days_and_hours_settings/show.html.erb:
```diff
@@ -27,22 +27,51 @@ See COPYRIGHT and LICENSE files for more details.
+    <fieldset class="form--fieldset">
+      <div class="form--field">
+        <%= setting_text_field :hours_per_day,
+                               type: :number %>
```
This field and the 2 others are larger than what was designed in figma. In the mockups they are only half width. You can achieve it by adding `container_class: "-wide"` argument. The css classes are defined in `frontend/src/global_styles/content/_forms.sass:579-598`

cbliard commented on app/views/admin/settings/working_days_and_hours_settings/show.html.erb:
```diff
@@ -27,22 +27,51 @@ See COPYRIGHT and LICENSE files for more details.
+        <%= t("working_days.warning").html_safe %>
+      </p>
+    </div>
+  </div>
```
I feel like this warning should belong to the next section. Currently it looks like it belongs to the wrong one.
Maybe @psatyal can confirm.

aaron-contreras replied:
You're right. Played around with it and like it more this way:

<img width="1300" alt="Banner position demo" src="https://github.com/opf/openproject/assets/61627014/7e4127b2-adbf-4584-bd13-e9bfd87ea7fd">

aaron-contreras replied:
Thoughts @psatyal ?

aaron-contreras replied:
This is the commit https://github.com/opf/openproject/pull/15450/commits/adfa1bd7f6863d2f953f25f95cbe9597b514e04d

in case we agree this is where it should be

cbliard replied:
Seen with Parimal during today's frontend meeting:
- the text has been slightly reworded
- the warning is moved on top of the section, so the warning can be read before doing the changes

![image](https://github.com/opf/openproject/assets/176055/d44c5148-0b41-4dac-912d-9021e66006a3)

He also asked if it's possible when clicking "Apply changes" to display the warning message about changed days only if there are actually changed days. I'll do another review to specifically ask for this so we can track it more easily.

cbliard commented on app/views/layouts/base.html.erb:
```diff
@@ -156,28 +156,28 @@ See COPYRIGHT and LICENSE files for more details.
-    var wrapper = $('#wrapper');
-    var savedMainMenuWidth = window.OpenProject.guardedLocalStorage("openProject-mainMenuWidth");
-    var mainMenuCollapsed =  window.OpenProject.guardedLocalStorage("openProject-mainMenuCollapsed");
+  var wrapper = $('#wrapper');
+  var savedMainMenuWidth = window.OpenProject.guardedLocalStorage("openProject-mainMenuWidth");
+  var mainMenuCollapsed = window.OpenProject.guardedLocalStorage("openProject-mainMenuCollapsed");
```
It looks like the indentation was all messed up in this function. Not sure why.

aaron-contreras replied:
Hmmm, I think it was a Rubymine "autocorrect-on-save".

aaron-contreras replied:
Fixed.

cbliard reviewed (commented):
I can't approve nor request changes on this PR because I created it 😅 

The multiple units display is really nice, and the way it's used in the popover is smooth and does not get in the way. That's nice!

I have some tiny remarks here and there.

I also have the feeling that the precision is annoying us. For instance:
* entering "2h 13m" leads to 7980 seconds, converted to 2.216666667 hours, rounded to 2.22 (2 decimals). This value 2.22 is the one being saved in database.
* When frontend is rendering it, it has 2 ways of rendering it:
  * the modal is rendering it through `DurationConverter.output`, which for 2.22 will render "2h 13m 12s"
  * the angular part is using a different route: the 2.22 is converted in the representer using `datetime_formatter.format_duration_from_hours`, so 2.22 is converted to iso duration "PT2H13M12S". This duration is then handled in the timezone service which converts it to hours with `Number(moment.duration(durationString).asHours().toFixed(2))` which is 2.22, converted to seconds which gives 7992.000000000001, and then it is converted to chronic duration which gives "2h 13m 12.000000000001s"

So the rounding is always false, and sometimes more false than others. I think the second one can be fixed by avoiding converting iso duration to hours and then to second: by converting iso duration directly to seconds, we could avoid precision loss.

Another thing we could do is to avoid displaying seconds entirely by making the hours to seconds conversion remove them. Something like `(hours * 3600 + 15).to_i / 60 * 60`. For 2.22 hours, it would give 7980 which converts to "2h 13m". No seconds, no errors.

WDYT?

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Rely on duration params being provided by parent class

aaron-contreras committed:
Ensure durations are positive numbers

aaron-contreras committed:
Limit the width of duration input fields

aaron-contreras committed:
Move toast warning to appropriate section

cbliard committed:
Move toast warning to even more appropriate location

Also:
- reword the warning message slightly to remove the "click apply to
  apply" part.
- remove some html_safe calls that are not needed anymore (especially
  since 2f82d73fb15).

cbliard commented on app/views/admin/settings/working_days_and_hours_settings/show.html.erb:
```diff
@@ -27,37 +27,66 @@ See COPYRIGHT and LICENSE files for more details.
+      <div class="form--field">
+        <%= setting_text_field :days_per_month,
+                               type: :number,
+                               container_class: "-wide",
+                               data: { admin__working_days_and_hours_target: 'daysPerMonthInput',
+                                       action: 'blur->admin--working-days-and-hours#recalculateDaysPerWeek' } %>
+        <div class="form--field-instructions"><%= t("setting_days_per_month_explanation") %></div>
+      </div>
+
+    </fieldset>
+  </section>
+
   <section class="form--section">
     <fieldset class="form--fieldset">
       <legend class="form--fieldset-legend"><%= t('settings.working_days.section_work_week') %></legend>
-
+      <div class="op-toast -warning">
+        <div class="op-toast--content">
+          <p>
+            <%= t("working_days.warning") %>
+          </p>
+        </div>
+      </div>
+      <p>
       <p>
-        <%= t("working_days.info").html_safe %>
+        <%= t("working_days.info") %>
       </p>
 
       <div class="form--field op-working-days-admin-settings--day-selectors" id="setting_working_days">
         <%= setting_multiselect :working_days,
-
-              I18n.t('date.day_names').rotate.zip(WeekDay::DAY_RANGE),
-              direction: :horizontal %>
+                                I18n.t('date.day_names').rotate.zip(WeekDay::DAY_RANGE),
+                                direction: :horizontal %>
```
When clicking "Apply changes", there is this "Change working days" information dialog that always opens itself to warn about changing working days.

![image](https://github.com/opf/openproject/assets/176055/be90913a-f2a0-497c-b44c-ab91f3abf29f)


Parimal would like it to be displayed only if the working days have been actually changed.

Would it be possible?

cbliard replied:
I'll create a bug work package for that one.

cbliard reviewed (commented):
An additional request from Parimal while discussing this screen during today's frontend meeting

aaron-contreras committed:
Assume next logical unit when parsing a string

aaron-contreras committed:
Remove seconds from output to prevent rounding errors

aaron-contreras committed:
Merge branch 'dev' into feature/50954-record-progress-in-multiple-units

aaron-contreras committed:
Ensure durations are properly validated

cbliard committed:
Fix text in days_per_week setting explanation

cbliard committed:
Fix tests due to changes in rounding duration to minute

cbliard reviewed (commented):
I can't approve because I created the PR, but I approve it 👍
